Dollar in identifier

When compiling with Clang and `-Weverything -Werror` we get this error:

```
/path/to/analysis.h:935:36: error: '$' in identifier [-Werror,-Wdollar-in-identifier-extension]
#ifndef CXXBRIDGE1_STRUCT_somenamespace$SomeStruct
```

Apparently whether `$` is allowed or not [is implementation defined](https://stackoverflow.com/a/26302259/265521). (C++ 🤷 )

It would be nice if `cxx` could use something else instead of `$` (maybe `____`?) or alternatively add pragmas to disable that warning in the generated code for Clang.

For now we have just disabled that specific warning in our build system.

--- a/gen/src/builtin.rs
+++ b/gen/src/builtin.rs
@@ -402,7 +402,11 @@ fn write_builtin<'a>(
             }
         } else if let Some(rest) = line.strip_prefix("#pragma GCC diagnostic ignored \"") {
             let diagnostic = rest.strip_suffix('"').unwrap();
-            pragma.diagnostic_ignore.insert(diagnostic);
+            pragma.gnu_diagnostic_ignore.insert(diagnostic);
+            ready = false;
+        } else if let Some(rest) = line.strip_prefix("#pragma clang diagnostic ignored \"") {
+            let diagnostic = rest.strip_suffix('"').unwrap();
+            pragma.clang_diagnostic_ignore.insert(diagnostic);
             ready = false;
         } else if line == "namespace {" {
             namespace.push(Block::AnonymousNamespace);

--- a/gen/src/builtin/trycatch_detail.h
+++ b/gen/src/builtin/trycatch_detail.h
@@ -3,6 +3,7 @@
 #include <string>
 
 #pragma GCC diagnostic ignored "-Wshadow"
+#pragma clang diagnostic ignored "-Wdollar-in-identifier-extension"
 
 namespace rust {
 inline namespace cxxbridge1 {

--- a/gen/src/guard.rs
+++ b/gen/src/guard.rs
@@ -1,0 +1,23 @@
+use crate::syntax::symbol::Symbol;
+use crate::syntax::Pair;
+use std::fmt::{self, Display};
+
+pub(crate) struct Guard {
+    kind: &'static str,
+    symbol: Symbol,
+}
+
+impl Guard {
+    pub fn new(kind: &'static str, name: &Pair) -> Self {
+        Guard {
+            kind,
+            symbol: name.to_symbol(),
+        }
+    }
+}
+
+impl Display for Guard {
+    fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        write!(formatter, "{}_{}", self.kind, self.symbol)
+    }
+}

--- a/gen/src/guard.rs
+++ b/gen/src/guard.rs
@@ -1,3 +1,4 @@
+use crate::gen::out::OutFile;
 use crate::syntax::symbol::Symbol;
 use crate::syntax::Pair;
 use std::fmt::{self, Display};
@@ -8,11 +9,10 @@ pub(crate) struct Guard {
 }
 
 impl Guard {
-    pub fn new(kind: &'static str, name: &Pair) -> Self {
-        Guard {
-            kind,
-            symbol: name.to_symbol(),
-        }
+    pub fn new(out: &mut OutFile, kind: &'static str, name: &Pair) -> Self {
+        let symbol = name.to_symbol();
+        out.pragma.dollar_in_identifier |= symbol.contains('$');
+        Guard { kind, symbol }
     }
 }
 

--- a/gen/src/mod.rs
+++ b/gen/src/mod.rs
@@ -8,6 +8,7 @@ mod check;
 pub(super) mod error;
 mod file;
 pub(super) mod fs;
+mod guard;
 mod ifndef;
 pub(super) mod include;
 mod names;

--- a/gen/src/pragma.rs
+++ b/gen/src/pragma.rs
@@ -3,7 +3,9 @@ use std::collections::BTreeSet;
 
 #[derive(Default)]
 pub(crate) struct Pragma<'a> {
-    pub diagnostic_ignore: BTreeSet<&'a str>,
+    pub gnu_diagnostic_ignore: BTreeSet<&'a str>,
+    pub clang_diagnostic_ignore: BTreeSet<&'a str>,
+    pub dollar_in_identifier: bool,
     pub begin: Content<'a>,
     pub end: Content<'a>,
 }
@@ -15,24 +17,46 @@ impl<'a> Pragma<'a> {
 }
 
 pub(super) fn write(out: &mut OutFile) {
-    if out.pragma.diagnostic_ignore.is_empty() {
-        return;
+    if out.pragma.dollar_in_identifier {
+        out.pragma
+            .clang_diagnostic_ignore
+            .insert("-Wdollar-in-identifier-extension");
     }
 
     let begin = &mut out.pragma.begin;
-    writeln!(begin, "#ifdef __GNUC__");
-    if out.header {
-        writeln!(begin, "#pragma GCC diagnostic push");
+    if !out.pragma.gnu_diagnostic_ignore.is_empty() {
+        writeln!(begin, "#ifdef __GNUC__");
+        if out.header {
+            writeln!(begin, "#pragma GCC diagnostic push");
+        }
+        for diag in &out.pragma.gnu_diagnostic_ignore {
+            writeln!(begin, "#pragma GCC diagnostic ignored \"{diag}\"");
+        }
     }
-    for diag in &out.pragma.diagnostic_ignore {
-        writeln!(begin, "#pragma GCC diagnostic ignored \"{diag}\"");
+    if !out.pragma.clang_diagnostic_ignore.is_empty() {
+        writeln!(begin, "#ifdef __clang__");
+        if out.header && out.pragma.gnu_diagnostic_ignore.is_empty() {
+            writeln!(begin, "#pragma clang diagnostic push");
+        }
+        for diag in &out.pragma.clang_diagnostic_ignore {
+            writeln!(begin, "#pragma clang diagnostic ignored \"{diag}\"");
+        }
+        writeln!(begin, "#endif // __clang__");
     }
-    writeln!(begin, "#endif");
+    if !out.pragma.gnu_diagnostic_ignore.is_empty() {
+        writeln!(begin, "#endif // __GNUC__");
+    }
 
     if out.header {
         let end = &mut out.pragma.end;
-        writeln!(end, "#ifdef __GNUC__");
-        writeln!(end, "#pragma GCC diagnostic pop");
-        writeln!(end, "#endif");
+        if !out.pragma.gnu_diagnostic_ignore.is_empty() {
+            writeln!(end, "#ifdef __GNUC__");
+            writeln!(end, "#pragma GCC diagnostic pop");
+            writeln!(end, "#endif // __GNUC__");
+        } else if !out.pragma.clang_diagnostic_ignore.is_empty() {
+            writeln!(end, "#ifdef __clang__");
+            writeln!(end, "#pragma clang diagnostic pop");
+            writeln!(end, "#endif // __clang__");
+        }
     }
 }

--- a/gen/src/write.rs
+++ b/gen/src/write.rs
@@ -1,4 +1,5 @@
 use crate::gen::block::Block;
+use crate::gen::guard::Guard;
 use crate::gen::nested::NamespaceEntries;
 use crate::gen::out::OutFile;
 use crate::gen::{builtin, include, pragma, Opt};
@@ -277,7 +278,7 @@ fn write_struct<'a>(out: &mut OutFile<'a>, strct: &'a Struct, methods: &[&Extern
     let operator_ord = derive::contains(&strct.derives, Trait::PartialOrd);
 
     out.set_namespace(&strct.name.namespace);
-    let guard = format!("CXXBRIDGE1_STRUCT_{}", strct.name.to_symbol());
+    let guard = Guard::new("CXXBRIDGE1_STRUCT", &strct.name);
     writeln!(out, "#ifndef {}", guard);
     writeln!(out, "#define {}", guard);
     write_doc(out, "", &strct.doc);
@@ -394,7 +395,7 @@ fn write_struct_using(out: &mut OutFile, ident: &Pair) {
 
 fn write_opaque_type<'a>(out: &mut OutFile<'a>, ety: &'a ExternType, methods: &[&ExternFn]) {
     out.set_namespace(&ety.name.namespace);
-    let guard = format!("CXXBRIDGE1_STRUCT_{}", ety.name.to_symbol());
+    let guard = Guard::new("CXXBRIDGE1_STRUCT", &ety.name);
     writeln!(out, "#ifndef {}", guard);
     writeln!(out, "#define {}", guard);
     write_doc(out, "", &ety.doc);
@@ -441,7 +442,7 @@ fn write_opaque_type<'a>(out: &mut OutFile<'a>, ety: &'a ExternType, methods: &[
 
 fn write_enum<'a>(out: &mut OutFile<'a>, enm: &'a Enum) {
     out.set_namespace(&enm.name.namespace);
-    let guard = format!("CXXBRIDGE1_ENUM_{}", enm.name.to_symbol());
+    let guard = Guard::new("CXXBRIDGE1_ENUM", &enm.name);
     writeln!(out, "#ifndef {}", guard);
     writeln!(out, "#define {}", guard);
     write_doc(out, "", &enm.doc);

--- a/gen/src/write.rs
+++ b/gen/src/write.rs
@@ -201,6 +201,7 @@ fn write_std_specializations(out: &mut OutFile, apis: &[Api]) {
                 out.next_section();
                 out.include.cstddef = true;
                 out.include.functional = true;
+                out.pragma.dollar_in_identifier = true;
                 let qualified = strct.name.to_fully_qualified();
                 writeln!(out, "template <> struct hash<{}> {{", qualified);
                 writeln!(
@@ -278,7 +279,7 @@ fn write_struct<'a>(out: &mut OutFile<'a>, strct: &'a Struct, methods: &[&Extern
     let operator_ord = derive::contains(&strct.derives, Trait::PartialOrd);
 
     out.set_namespace(&strct.name.namespace);
-    let guard = Guard::new("CXXBRIDGE1_STRUCT", &strct.name);
+    let guard = Guard::new(out, "CXXBRIDGE1_STRUCT", &strct.name);
     writeln!(out, "#ifndef {}", guard);
     writeln!(out, "#define {}", guard);
     write_doc(out, "", &strct.doc);
@@ -395,7 +396,7 @@ fn write_struct_using(out: &mut OutFile, ident: &Pair) {
 
 fn write_opaque_type<'a>(out: &mut OutFile<'a>, ety: &'a ExternType, methods: &[&ExternFn]) {
     out.set_namespace(&ety.name.namespace);
-    let guard = Guard::new("CXXBRIDGE1_STRUCT", &ety.name);
+    let guard = Guard::new(out, "CXXBRIDGE1_STRUCT", &ety.name);
     writeln!(out, "#ifndef {}", guard);
     writeln!(out, "#define {}", guard);
     write_doc(out, "", &ety.doc);
@@ -442,7 +443,7 @@ fn write_opaque_type<'a>(out: &mut OutFile<'a>, ety: &'a ExternType, methods: &[
 
 fn write_enum<'a>(out: &mut OutFile<'a>, enm: &'a Enum) {
     out.set_namespace(&enm.name.namespace);
-    let guard = Guard::new("CXXBRIDGE1_ENUM", &enm.name);
+    let guard = Guard::new(out, "CXXBRIDGE1_ENUM", &enm.name);
     writeln!(out, "#ifndef {}", guard);
     writeln!(out, "#define {}", guard);
     write_doc(out, "", &enm.doc);
@@ -548,6 +549,7 @@ fn write_struct_operator_decls<'a>(out: &mut OutFile<'a>, strct: &'a Struct) {
     out.begin_block(Block::ExternC);
 
     if derive::contains(&strct.derives, Trait::PartialEq) {
+        out.pragma.dollar_in_identifier = true;
         let link_name = mangle::operator(&strct.name, "eq");
         writeln!(
             out,
@@ -566,6 +568,7 @@ fn write_struct_operator_decls<'a>(out: &mut OutFile<'a>, strct: &'a Struct) {
     }
 
     if derive::contains(&strct.derives, Trait::PartialOrd) {
+        out.pragma.dollar_in_identifier = true;
         let link_name = mangle::operator(&strct.name, "lt");
         writeln!(
             out,
@@ -599,6 +602,7 @@ fn write_struct_operator_decls<'a>(out: &mut OutFile<'a>, strct: &'a Struct) {
 
     if derive::contains(&strct.derives, Trait::Hash) {
         out.include.cstddef = true;
+        out.pragma.dollar_in_identifier = true;
         let link_name = mangle::operator(&strct.name, "hash");
         writeln!(
             out,
@@ -618,6 +622,8 @@ fn write_struct_operators<'a>(out: &mut OutFile<'a>, strct: &'a Struct) {
     out.set_namespace(&strct.name.namespace);
 
     if derive::contains(&strct.derives, Trait::PartialEq) {
+        out.pragma.dollar_in_identifier = true;
+
         out.next_section();
         writeln!(
             out,
@@ -644,6 +650,8 @@ fn write_struct_operators<'a>(out: &mut OutFile<'a>, strct: &'a Struct) {
     }
 
     if derive::contains(&strct.derives, Trait::PartialOrd) {
+        out.pragma.dollar_in_identifier = true;
+
         out.next_section();
         writeln!(
             out,
@@ -697,6 +705,7 @@ fn write_struct_operators<'a>(out: &mut OutFile<'a>, strct: &'a Struct) {
 fn write_opaque_type_layout_decls<'a>(out: &mut OutFile<'a>, ety: &'a ExternType) {
     out.set_namespace(&ety.name.namespace);
     out.begin_block(Block::ExternC);
+    out.pragma.dollar_in_identifier = true;
 
     let link_name = mangle::operator(&ety.name, "sizeof");
     writeln!(out, "::std::size_t {}() noexcept;", link_name);
@@ -713,6 +722,7 @@ fn write_opaque_type_layout<'a>(out: &mut OutFile<'a>, ety: &'a ExternType) {
     }
 
     out.set_namespace(&ety.name.namespace);
+    out.pragma.dollar_in_identifier = true;
 
     out.next_section();
     let link_name = mangle::operator(&ety.name, "sizeof");
@@ -742,6 +752,7 @@ fn begin_function_definition(out: &mut OutFile) {
 }
 
 fn write_cxx_function_shim<'a>(out: &mut OutFile<'a>, efn: &'a ExternFn) {
+    out.pragma.dollar_in_identifier = true;
     out.next_section();
     out.set_namespace(&efn.name.namespace);
     out.begin_block(Block::ExternC);
@@ -954,6 +965,7 @@ fn write_rust_function_decl_impl(
     indirect_call: bool,
 ) {
     out.next_section();
+    out.pragma.dollar_in_identifier = true;
     if sig.throws {
         out.builtin.ptr_len = true;
         write!(out, "::rust::repr::PtrLen ");
@@ -1088,6 +1100,7 @@ fn write_rust_function_shim_impl(
         // We've already defined this inside the struct.
         return;
     }
+    out.pragma.dollar_in_identifier = true;
     if matches!(sig.kind, FnKind::Free) {
         // Member functions already documented at their declaration.
         write_doc(out, "", doc);
@@ -1513,6 +1526,8 @@ fn write_rust_box_extern(out: &mut OutFile, key: &NamedImplKey) {
     let inner = resolve.name.to_fully_qualified();
     let instance = resolve.name.to_symbol();
 
+    out.pragma.dollar_in_identifier = true;
+
     writeln!(
         out,
         "{} *cxxbridge1$box${}$alloc() noexcept;",
@@ -1536,6 +1551,7 @@ fn write_rust_vec_extern(out: &mut OutFile, key: &NamedImplKey) {
     let instance = element.to_mangled(out.types);
 
     out.include.cstddef = true;
+    out.pragma.dollar_in_identifier = true;
 
     writeln!(
         out,
@@ -1584,6 +1600,8 @@ fn write_rust_box_impl(out: &mut OutFile, key: &NamedImplKey) {
     let inner = resolve.name.to_fully_qualified();
     let instance = resolve.name.to_symbol();
 
+    out.pragma.dollar_in_identifier = true;
+
     writeln!(out, "template <>");
     begin_function_definition(out);
     writeln!(
@@ -1617,6 +1635,7 @@ fn write_rust_vec_impl(out: &mut OutFile, key: &NamedImplKey) {
     let instance = element.to_mangled(out.types);
 
     out.include.cstddef = true;
+    out.pragma.dollar_in_identifier = true;
 
     writeln!(out, "template <>");
     begin_function_definition(out);
@@ -1708,6 +1727,8 @@ fn write_unique_ptr(out: &mut OutFile, key: &NamedImplKey) {
 fn write_unique_ptr_common(out: &mut OutFile, ty: UniquePtr) {
     out.include.new = true;
     out.include.utility = true;
+    out.pragma.dollar_in_identifier = true;
+
     let inner = ty.to_typename(out.types);
     let instance = ty.to_mangled(out.types);
 
@@ -1814,6 +1835,7 @@ fn write_shared_ptr(out: &mut OutFile, key: &NamedImplKey) {
 
     out.include.new = true;
     out.include.utility = true;
+    out.pragma.dollar_in_identifier = true;
 
     // Some aliases are to opaque types; some are to trivial types. We can't
     // know at code generation time, so we generate both C++ and Rust side
@@ -1909,6 +1931,7 @@ fn write_weak_ptr(out: &mut OutFile, key: &NamedImplKey) {
 
     out.include.new = true;
     out.include.utility = true;
+    out.pragma.dollar_in_identifier = true;
 
     writeln!(
         out,
@@ -1979,6 +2002,7 @@ fn write_cxx_vector(out: &mut OutFile, key: &NamedImplKey) {
     out.include.cstddef = true;
     out.include.utility = true;
     out.builtin.destroy = true;
+    out.pragma.dollar_in_identifier = true;
 
     begin_function_definition(out);
     writeln!(

--- a/src/cxx.cc
+++ b/src/cxx.cc
@@ -25,6 +25,9 @@
 #ifdef __GNUC__
 #pragma GCC diagnostic ignored "-Wshadow"
 #endif
+#ifdef __clang__
+#pragma clang diagnostic ignored "-Wdollar-in-identifier-extension"
+#endif
 
 extern "C" {
 void cxxbridge1$cxx_string$init(std::string *s, const std::uint8_t *ptr,

--- a/syntax/symbol.rs
+++ b/syntax/symbol.rs
@@ -38,6 +38,11 @@ impl Symbol {
         assert!(!symbol.0.is_empty());
         symbol
     }
+
+    #[allow(dead_code)]
+    pub(crate) fn contains(&self, ch: char) -> bool {
+        self.0.contains(ch)
+    }
 }
 
 pub(crate) trait Segment {

